woocommerce/woocommerce
# Logging: New, improved view for a single log file

### Changes proposed in this Pull Request:

This provides a new view of a log file when you click on one in the https://github.com/woocommerce/woocommerce/pull/40662. It has several enhancements over the original log file view:

* Clickable line numbers. These provide permalinks to specific lines in the file.
* Formatting of some log entry elements. The severity level gets colorized so it's easier to see which log entries are more important, and the timestamp gets bolded so you can see where each log entry starts (since some entries are across multiple lines).
* Navigation links for switching between different "rotations" of a particular log.

This also refactors the File class a bit to use `WP_Filesystem` wherever possible instead of direct PHP filesystem calls.

Fixes #41143 

![Screen Shot 2023-11-02 at 16 08 37](https://github.com/woocommerce/woocommerce/assets/916023/c9629df7-826d-4b96-98fd-351d2d6a24b2)

### How to test the changes in this Pull Request:

#### Setup

1. In your test site's **wp-config.php** file, add this line: `define( 'WC_LOG_HANDLER', 'Automattic\\WooCommerce\\Internal\\Admin\\Logging\\LogHandlerFileV2' );`. This is what allows you to see the new log file views.
2. Install [this plugin](https://gist.github.com/coreymckrill/2968911c8b63c556e7787f2cdc65f966) on your test site for generating test log files. Probably easiest to copy the file into the mu-plugins directory.
3. Open two browser tabs: the Logs screen (WP Admin > WooCommerce > Status > Logs) and the WooCommerce Tools screen (WP Admin > WooCommerce > Status > Tools). On the Tools screen you should see the **Debug Log Generator** tool at the top of the list.

#### Tests

1. On the Logs screen, you should see a list table, which may already have some log files listed in it. Even if there are some, go ahead and generate some additional logs on the Tools screen so that we have enough different log files and sources. Generate two or three batches of test logs at once. You should then see that at least one of the logs, probably `debug-log-generator`, has more than one entry in the list table. These are "rotations" that are generated when a log file reaches a certain size.
2. Click on a log file that has one or more rotations. This should take you to the single file view, where you can see the contents of the file. There should be line numbers next to each line of the file.
3. Try clicking the line numbers. They are anchors, and the line you clicked should move to the top of the window (but not beneath the floating toolbars), assuming there are enough lines in the file that you can't see all of them at once. The line should also be highlighted in yellow.
4. Since this is a file that has rotations, you should see a label that says "File rotations" and some buttons next to it. Possible button labels are "Current" as well as any single-digit number between `0` and `9`. In this case "Current" means the file that has not hit the size threshold yet, so new log entries could be added to it. The numbered ones are files that have been rotated, with lower numbers being more recent than higher numbers. Clicking one of the buttons should take you to that file.
5. Try deleting one of the log files by clicking the "Delete permanently" button in the upper right corner. You should see a confirmation dialog popup before the file gets deleted. If you click "Cancel", the file should not be deleted!

### Changelog entry




<details>

#### Significance




#### Type




#### Message

Add a new single file view to the Logs screen for viewing the contents of a log file.

#### Comment 

</details>
